packages/wallet-protocols/src/protocols/virtual-fund-as-leaf/protocol.ts

```diff
@@ -1,131 +1,167 @@
+    destinations: [participants[2 * index], participants[1]].map(makeDestination),
   };
-}
 
-const createGuarantorChannel = {
-  invoke: {
-    src: 'createNullChannel',
-    data: 'guarantorChannelArgs',
-  },
-};
-
-function fundGuarantorArgs({ guarantorChannel, ledgerId, balances }: ChannelsKnown) {
-  const amount = total(balances);
   return {
-    channelId: ledgerId,
-    outcome: ethAllocationOutcome(
-      [{ destination: getChannelId(guarantorChannel), amount }],
-      AddressZero /* TODO: Should be store.ethassetholderaddress */
-    ),
+    channel: guarantorChannel,
+    outcome: ethGuaranteeOutcome(guarantee, store.ethAssetHolderAddress),
   };
-}
+};
+const createGuarantorChannel = getDataAndInvoke(
+  'guarantorChannelArgs',
+  'createNullChannel',
+  undefined,
+  'createGuarantorChannel'
+);
+
 const createChannels = {
-  entry: 'assignChannels',
-  type: 'parallel',
-  states: {
-    createGuarantorChannel,
-    createJointChannel,
-  },
+  type: 'parallel' as 'parallel',
+  states: { createGuarantorChannel, createJointChannel },
   onDone: 'fundGuarantor',
 };
 
-const fundGuarantor = {
-  invoke: {
-    src: 'supportState',
-    data: 'guarantorOutcome',
-    onDone: 'fundTarget',
-  },
+export const fundGuarantorArgs = ({
+  guarantorChannel,
+  balances,
+  index,
+  hubAddress,
+}: Pick<Init, 'guarantorChannel' | 'balances' | 'hubAddress' | 'index'>): LedgerFunding.Init => {
+  const deductions: Allocation = balances.map(allocationItem);
+  deductions[1 - index].destination = makeDestination(hubAddress);
+
+  return { targetChannelId: getChannelId(guarantorChannel), deductions };
 };
 
-const fundTarget = {
-  invoke: {
-    src: 'supportState',
-    data: 'jointOutcome',
-    onDone: 'success',
-  },
+const fundGuarantor = {
+  invoke: { src: 'ledgerFunding', data: fundGuarantorArgs, onDone: 'fundTarget' },
 };
 
+export function fundTargetArgs(store: Store) {
+  return async ({
+    jointChannel,
+    balances,
+    hubAddress,
+    targetChannelId,
+  }: Init): Promise<SupportState.Init> => {
+    const { latestSupportedState } = await store.getEntry(getChannelId(jointChannel));
+    const expectedAllocation = jointChannelAllocation(balances, hubAddress);
+    if (
+      !outcomesEqual(
+        latestSupportedState.outcome,
+        ethAllocationOutcome(expectedAllocation, store.ethAssetHolderAddress)
+      )
+    ) {
+      throw 'Unexpected outcome';
+    }
+
+    const { amount } = expectedAllocation[1];
+    return {
+      state: {
+        ...latestSupportedState,
+        turnNum: latestSupportedState.turnNum + 1,
+        outcome: ethAllocationOutcome(
+          [
+            { destination: targetChannelId, amount },
+            { destination: makeDestination(hubAddress), amount },
+          ],
+          store.ethAssetHolderAddress
+        ),
+      },
+    };
+  };
+}
+const fundTarget = getDataAndInvoke('fundTargetArgs', 'supportState', 'success');
```

Comment:
This isn't really a comment about this specific PR, but more about x-state, or at least the way we're using it. This line makes me feel really uneasy. As someone who's reading the code, essentially there are some magic strings being passed to a function. There doesn't seem to be an easy way for me to trace these back to where they're defined, or inspect their types, or understand what's going to be passed to them. I get that connecting services to machine config via strings allows for the implementation to be decoupled, but it seems like we have so much decoupling that it becomes harder to figure out what the actual code does.

It would be good to discuss this further, and figure out if there are patterns we can use to make our x-state code easier to read and understand.